sig-scalability/slos/dns_latency.md

```diff
@@ -0,0 +1,44 @@
+## In-cluster dns latency SLIs/SLOs details
+
+### Definition
+
+| Status | SLI | SLO |
+| --- | --- | --- |
+| __WIP__ | In-cluster dns latency from a single prober pod, measured as latency of per second DNS lookup for "null service" from that pod, measured as 99th percentile over last 5 minutes. | In default Kubernetes installataion with RTT between nodes <= Y, 99th percentile of (99th percentile over all prober pods) per cluster-day <= X |
```

Comment:
Do we care about node-local caching having an impact on this?  Maybe we actually want 2 metrics ?

Reply:
I was thinking about that. But do we have a way to differentiate those?

Reply:
We could have one set of prober pods use the cluster-dns service ip and another set use the local dns cache ip. That way we can differenciate the 2?

Reply:
local caching would have to be taken into account when querying for a "null service" and negative caching is enabled

Reply:
I assumed a "null" service is something like `kubernetes.default`.

@prameshj do we have a way for user to express that they don't want the cache?  I guess the prober could simply look up the IP of `kube-dns.kube-system` and use that, but I am a little wary of that, for some reason.  Imagine a hosted cluster that ran DNS not as a pod but a managed service.  I guess we could also demand that Service be created.

e.g.

1) get the nameserver IP from `/etc/resolv.conf` consider that "default"
2) if service kube-system/kube-dns exists consider that "upstream"

if "upstream" != "default", probe both.  Else just probe "default".

Reply:
Currently, the local cache is disabled by default. User needs to set an environment variable at cluster creation to enable it.
Ideally we want 2 clusters created - one with the cache enabled and one without. Then just use the pod's /etc/resolv.conf.

There are scripts to configure the local cache such that it listens on the existing kube-dns ip - https://github.com/kubernetes/dns/pull/280
It is not the default behavior, but relying on ip address to figure out if the local cache is running might not always work.

Reply:
We can't really say that "we want 2 clusters created". We would like to be able to measure SLIs in production clusters too (where we can't say: "create your cluster this way"). Yes - it requires running probers to measure that, but we would like to deliver those so if someone would like to measure it, they would be able to use it.

So I'm still not sure what we should do about that.
Maybe what we should say is:
- we always probe whatever is in pods /etc/resolv.conf
- we admit that depending on your cluster setup, it may or may not use local-cache, so the results may significantly vary.

Does that sound reasonable?

Reply:
Why is what I wrote above not feasible?  If we can detect 2 levels of DNS we can probe and measure them both.  We don't need to SLO both, but maybe we want to.  

Alternatively, we could do something like probe for local names that are likely to be cached (e.g. kubernetes.default) and names that are local and unlikely to be cached ($random.$random) and maybe even probe for names that are unlikely to be local and unlikely to be cached (e.g. $random.gke.io which we could even answer with a wildcard or $random.internal (via nanny))

Reply:
Ohh sorry - I was kind of referring to @prameshj at the end of his comment "but relying on ip address to figure out if the local cache is running might not always work."
But maybe I misinterpreted it?

I'm fine with probing two address as you suggested (I mean the address from /etc/resolv.conf and kube-dns service), if you and others believe that's what we should do (it just wasn't clear for me).

Reply:
Yes, that is what I was saying in my previous comment. If we look at /etc/resolv.conf and find that to be same as the kube-dns service ip, it could still be the case that we have a local cache that is using the same ip.
It is not the design today for nodelocal cache, but it could change. We can use the 2 address approach for figuring out what the numbers are like. 

I like the idea of picking some hostnames that are likely to be cached and some that are not.

Reply:
I tried to capture that in the SLI definition (or actually a footnote I added to it).
PTAL